server/src/main/java/org/elasticsearch/index/mapper/DocCountFieldMapper.java

```diff
@@ -0,0 +1,159 @@
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.index.mapper;
+
+import org.apache.lucene.document.Field;
+import org.apache.lucene.document.FieldType;
+import org.apache.lucene.document.NumericDocValuesField;
+import org.apache.lucene.index.DocValuesType;
+import org.apache.lucene.index.IndexOptions;
+import org.apache.lucene.search.DocValuesFieldExistsQuery;
+import org.apache.lucene.search.Query;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.index.query.QueryShardContext;
+import org.elasticsearch.index.query.QueryShardException;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import static org.elasticsearch.index.mapper.TypeParsers.parseField;
+
+/** Mapper for the doc_count field. */
+public class DocCountFieldMapper extends FieldMapper {
+
+    public static final String CONTENT_TYPE = "doc_count";
+
+    public static class Defaults {
+        public static final FieldType FIELD_TYPE = new FieldType();
+
+        static {
+            FIELD_TYPE.setDocValuesType(DocValuesType.NUMERIC);
+            FIELD_TYPE.setIndexOptions(IndexOptions.NONE);
+            FIELD_TYPE.freeze();
+        }
+    }
+
+    public static class Builder extends FieldMapper.Builder<DocCountFieldMapper.Builder> {
+
+        public Builder(String name) {
+            super(name, Defaults.FIELD_TYPE);
+            builder = this;
+        }
+
+        @Override
+        public DocCountFieldMapper build(BuilderContext context) {
+            DocCountFieldType defaultFieldType = new DocCountFieldType(buildFullName(context), hasDocValues, meta);
+            return new DocCountFieldMapper(name, fieldType, defaultFieldType);
+        }
+    }
+
+    public static class TypeParser implements Mapper.TypeParser {
+
+        public TypeParser() {
+        }
+
+        @Override
+        public DocCountFieldMapper.Builder parse(String name, Map<String, Object> node, ParserContext parserContext)
+            throws MapperParsingException {
+            DocCountFieldMapper.Builder builder = new DocCountFieldMapper.Builder(name);
+            parseField(builder, name, node, parserContext);
+            return builder;
+        }
+    }
+
+    public static final class DocCountFieldType extends MappedFieldType {
+
+        public DocCountFieldType(String name, boolean hasDocValues, Map<String, String> meta) {
+            super(name, false, hasDocValues, TextSearchInfo.SIMPLE_MATCH_ONLY, meta);
```

Comment:
Needs verification from someone that knows mappers better than me, but I think if we set `IndexOptions.NONE` we should probably also set `TextSearchInfo.NONE`.  I think?  Not entirely confident in that

Reply:
Class `VersionFieldMapper` uses similar pattern (`IndexOptions.NONE` and `TextSearchInfo.NONE`)

https://github.com/elastic/elasticsearch/blob/62f51eb9aec6c3d423d7fa9645ebe9ef2904d4c1/server/src/main/java/org/elasticsearch/index/mapper/VersionFieldMapper.java#L75

I am not an expert on this either. 

@jtibshirani could you help with this please?

Reply:
I also think this should be `TextSearchInfo.NONE`, given we only allow `IndexOptions.NONE`. The choice in `VersionFieldMapper` seems like a mistake -- `TextSearchInfo` was created recently and we probably this during review.

Reply:
Sure, happy to change that to `TextSearchInfo.NONE`